node/pkg/driver/device_connectivity/device_connectivity_fc.go

```diff
@@ -0,0 +1,324 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package device_connectivity
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"errors"
+
+	"github.com/ibm/ibm-block-csi-driver/node/logger"
+	"github.com/ibm/ibm-block-csi-driver/node/pkg/driver/executer"
+)
+
+type OsDeviceConnectivityFc struct {
+	Executer          executer.ExecuterInterface
+	Helper            OsDeviceConnectivityHelperFcInterface
+	HelperScsiGeneric OsDeviceConnectivityHelperScsiGenericInterface
+}
+
+func NewOsDeviceConnectivityFc(executer executer.ExecuterInterface) OsDeviceConnectivityInterface {
+	return &OsDeviceConnectivityFc{
+		Executer:          executer,
+		Helper:            NewOsDeviceConnectivityHelperFc(executer),
+		HelperScsiGeneric: NewOsDeviceConnectivityHelperScsiGeneric(executer),
+	}
+}
+
+func (r OsDeviceConnectivityFc) RescanDevices(lunId int, arrayIdentifiers []string) error {
+	logger.Debugf("Rescan : Start rescan on specific lun, on lun : {%v}, with array wwn : {%v}", lunId, arrayIdentifiers)
+
+	if len(arrayIdentifiers) == 0 {
+		e := &ErrorNotFoundArrayIdentifiers{lunId}
+		logger.Errorf(e.Error())
+		return e
+	}
+
+	HostIDs, err := r.Helper.GetFcHostIDs()
+	if err != nil {
+		return err
+	}
+
+	return r.HelperScsiGeneric.RescanDevices(lunId, arrayIdentifiers, HostIDs)
+}
+
+func (r OsDeviceConnectivityFc) GetMpathDevice(volumeId string, lunId int, arrayIdentifiers []string) (string, error) {
+	/*
+			   Description:
+				   1. Find all the files "/dev/disk/by-path/pci-*-fc-<WWN storage>-lun-<LUN-ID> -> ../../sd<X>
+		 	          Note: <ARRAY-IP> Instead of setting here the IP we just search for * on that.
+		 	       2. Get the sd<X> devices.
+		 	       3. Search '/sys/block/dm-*\/slaves/*' and get the <DM device name>. For example dm-3 below:
+		 	          /sys/block/dm-3/slaves/sdb -> ../../../../pci0000:00/0000:00:17.0/0000:13:00.0/host33/rport-33:0-3/target33:0:1/33:0:1:0/block/sdb
+
+			   Return Value: "dm-X" of the volumeID by using the LunID and the array wwn.
+	*/
+	var devicePaths []string
+	var errStrings []string
+	lunIdStr := strconv.Itoa(lunId)
+
+	if len(arrayIdentifiers) == 0 {
+		e := &ErrorNotFoundArrayIdentifiers{lunId}
+		return "", e
+	}
+
+	for _, arrayIdentifier := range arrayIdentifiers {
+		wwn := "0x" + arrayIdentifier
+		dp := strings.Join([]string{"/dev/disk/by-path/pci*", "fc", wwn, "lun", lunIdStr}, "-")
+		logger.Infof("GetMpathDevice: Get the mpath devices related to wwn=%s and lunID=%s : {%v}", wwn, lunIdStr, dp)
+		dps, exists, e := r.Helper.WaitForPathToExist(dp, 5, 1)
+		if e != nil {
+			logger.Errorf("GetMpathDevice: No device found error : %v ", e.Error())
+			errStrings = append(errStrings, e.Error())
+		} else if !exists {
+			e := &MultipleDeviceNotFoundForLunError{volumeId, lunId, []string{wwn}}
+			logger.Errorf(e.Error())
+			errStrings = append(errStrings, e.Error())
```

Comment:
As mentioned in PR: https://github.com/IBM/ibm-block-csi-driver/pull/59
apimachinery has a very nice wrapper for a slice of errors in "k8s.io/apimachinery/pkg/util/errors", you can use errors.NewAggregate(errs) to convert a slice of errors into an aggregate error.

Reply:
xiaoyan working on it, I will refactor it after her finished it